### PR TITLE
PE-464: download-file 

### PR DESCRIPTION
**Pull request by matibat**

### CODE REVIEW

File: src/commands/download_file.ts
Comment (matibat) [inline src/commands/download_file.ts:22]: What if I provide the drive key as parameter? Drive ID shouldn't be required at all

Comment (fedellen) [reply src/commands/download_file.ts:22]: We should probably just eat the `getDriveIdForFileId` GQL call rather than having a driveID param. That is the pattern that most of the other commands are using.

You do bring up a great point here in that case too though, we wouldn't even need the drive ID if they supplied a driveKey. That inefficiency shows up in other commands where we don't pass down a driveId too, like `file-info` for example.

**Comment by fedellen** [comment]
We should have a README component here. At the very least we can add it to the command list at the bottom and maybe add a download section later.

**Comment by fedellen** [comment]
I ran a quick dev test with a local Core @ `PE-464_download_file`. The file seems to have successfully downloaded but I also ran into this error here:

```shell
 yarn ardrive download-file -f ff450770-a9cb-46a5-9234-89cbd9796610 -d bc9af866-6421-40f1-ac89-202bddb5c487                      File with ID "ff450770-a9cb-46a5-9234-89cbd9796610" was successfully download to "/Users/derek/ardrive/ardrive-cli"
(node:8109) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'fileEntity' of undefined
    at setLastModifiedDate (/Users/derek/ardrive/ardrive-cli/.yarn/unplugged/ardrive-core-js-file-4af702b991/node_modules/ardrive-core-js/lib/arfs/arfs_file_wrapper.js:228:60)
    at <anonymous>
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:8109) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:8109) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

File: package.json
Comment (arielmelendez) [inline package.json:11]: TODO: Update

File: src/commands/download_file.ts
Comment (arielmelendez) [reply src/commands/download_file.ts:22]: I agree. Let's make driveKey and driveID optional for this command if possible.

File: src/commands/download_file.ts
Comment (arielmelendez) [inline src/commands/download_file.ts:29]: What do you guys think of this behavior for -l:
• if the path is an existing folder path on disk, deposit the file in that folder based on its current name on ArDrive
• if the path is NOT an existing folder path on disk, assume the user wants to use it as the local file path, ignoring the file's name on ArDrive. To Discuss: Should we create any needed folder paths on the way to the file if they don't exist?

**Review by arielmelendez — CHANGES_REQUESTED** [review]
See my comments

File: package.json
Comment (matibat) [reply package.json:11]: That core branch is the right one

File: src/commands/download_file.ts
Comment (matibat) [reply src/commands/download_file.ts]: This was already applied :+1:

File: src/commands/download_file.ts
Comment (fedellen) [inline src/commands/download_file.ts:27]: This is still required at the parameter declaration level

Comment (fedellen) [reply src/commands/download_file.ts:27]: Even for public downloads

**Comment by fedellen** [comment]
Downloading a 200MB private file I ran into this error:

```shell
Error: Premature close
```

But it only happened on the first run, haven't been able to reproduce it to print the stack trace. It worked on the next 5 attempts

Might've been an environment issue, but leaving this here in case you have seen this before

File: src/commands/download_file.ts
Comment (fedellen) [reply src/commands/download_file.ts:30]: That behavior might feel more natural to a user. 

Right now it seems to be only used as the folder path. If we keep this current behavior for now, I think we should at least use a new parameter with the name DestinationFolderPath or similar so that users don't confuse it with a file name parameter.

>  To Discuss: Should we create any needed folder paths on the way to the file if they don't exist?

I think yes. Creating folders for the user provides convenience and should be arbitrary with `fs` or similar

File: src/commands/download_file.ts
Comment (fedellen) [inline src/commands/download_file.ts:15]: Make this non-required here or remove the DriveID param altogether

```suggestion
	parameters: [
		FileIdParameter,
		LocalFilePathParameter,
		{ name: DriveIdParameter, required: false },
		...DrivePrivacyParameters
	],
```

Comment (fedellen) [reply src/commands/download_file.ts:15]: And probably switch `LocalFilePathParameter` for a new `DestFolderPathParameter` since that is the behavior its currently using

**Review by fedellen — CHANGES_REQUESTED** [review]
A couple of small things. DriveID still seems to be required and the help output for the `-l` param here could be confusing to a user:

```shell
  -l --local-file-path <localFilePath>              the path on the local filesystem for the file that will be uploaded
```

File: src/commands/download_file.ts
Comment (matibat) [reply src/commands/download_file.ts:35]: > To Discuss: Should we create any needed folder paths on the way to the file if they don't exist?

I'd rather add a flag to force creating the directory tree the same way `mkdir -p` does. If the parent folder does not exist and no recursive flag provided, then throw the default `ENOENT`

File: src/commands/download_file.ts
Comment (matibat) [reply src/commands/download_file.ts]: 95e614a

Comment (matibat) [reply src/commands/download_file.ts]: d86bb4f7a9784e34c2bd40092b72c0e9cc07886b

Comment (matibat) [reply src/commands/download_file.ts]: Removed

File: src/commands/download_file.ts
Comment (matibat) [reply src/commands/download_file.ts:23]: See the commit https://github.com/ardriveapp/ardrive-core-js/pull/91/commits/2412b211c28d722c3b99d392d644631773cdd872 at core for the localFilepathResolution algorithm

File: README.md
Comment (matibat) [inline README.md:1008]: TODO: add doc

**Review by fedellen — DISMISSED** [review]
👍

File: package.json
Comment (arielmelendez) [reply package.json:11]: Thanks, but the reminder is just to update this back to the correct npm release before merging :)

File: src/commands/download_file.ts
Comment (arielmelendez) [reply src/commands/download_file.ts:23]: > I'd rather add a flag to force creating the directory tree the same way mkdir -p does. If the parent folder does not exist and no recursive flag provided, then throw the default ENOENT

Let's make a decision based on the ease of use of the following permutations:
1) -l path is a folder on the local file system that exists
    • Result: CLI uses folder as destPath and provides an undefined filename, leading to the ArDrive filename being used in the full path.
2) -l path is a file on the local file system that exists
    • Result: CLI uses the parent folder of the file as destPath and provides the filename. ArDrive filename is ignored.
3) -l path is a nonexistent entity on the local file system, but whose parent folder exists
    • Result: CLI uses the parent folder of the entity as destPath and provides the entity name as a filename. ArDrive filename is ignored. (DOES THIS SEEM CORRECT?!?)
3) -l path is a nonexistent entity on the local file system whose parent folder does NOT exist
    • Result: ERROR

Let's see how a flag like `mkdir`'s `-p` would change these cases:
1) No change
2) No change
3) No change, DEPENDING on whether we think the behavior described above is desirable
4) The path to the parent folder of the entity is created. The entity name is used. ArDrive name is ignored.

@agsuy should weigh in here.

File: src/commands/download_file.ts
Comment (matibat) [inline src/commands/download_file.ts:16]: Create a type for the returned arr

File: src/parameter_declarations.ts
Comment (matibat) [inline src/parameter_declarations.ts:238]: Rephrase this

File: src/utils/local_file_path.ts
Comment (matibat) [inline src/utils/local_file_path.ts:4]: Add docblock & set the return type

File: src/commands/download_file.ts
Comment (arielmelendez) [reply src/commands/download_file.ts]: Ideas for the flag, if we ever use it:
--create-paths
--parents
--create-intermediates alias: -i

File: src/parameter_declarations.ts
Comment (arielmelendez) [reply src/parameter_declarations.ts:238]:
```suggestion
	description: `a destination path for the file. Uses same destination rules as cp, i.e.:
\t\t\t\t\t\t\t-l /path/to/existing/folder (on-chain filename used in existing folder)
\t\t\t\t\t\t\t-l /path/to/existing/file (on-chain filename ignored, local file overwritten)
\t\t\t\t\t\t\t-l /path/to/existing/folder/nonexistent_thing (nonexistent_thing used as filename)
\t\t\t\t\t\t\t-l /path/to/nonexistent/folder (error)
`,
```

File: src/utils/local_file_path.ts
Comment (matibat) [reply src/utils/local_file_path.ts]: 223cd4e328a496788dff91737832bd612ab359ee

File: src/commands/download_file.ts
Comment (matibat) [reply src/commands/download_file.ts]: 223cd4e328a496788dff91737832bd612ab359ee

File: README.md
Comment (matibat) [reply README.md:1023]: 19a4cee745be160ec1525fe08aac66c4e003e23f

**Comment by matibat** [comment]
Re-updated the readme. Thanks for your suggestions @arielmelendez

**Review by fedellen — APPROVED** [review]
Working well on my end!

**Review by arielmelendez — APPROVED** [review]

**Review by fedellen — APPROVED** [review]

File: src/parameter_declarations.ts
Comment (matibat) [inline src/parameter_declarations.ts:23]: `--local-path` !

File: src/parameter_declarations.ts
Comment (arielmelendez) [inline src/parameter_declarations.ts:239]: I don't think anything from here down matters.

Comment (matibat) [reply src/parameter_declarations.ts]: f9171f2

**Review by arielmelendez — APPROVED** [review]

**Review by arielmelendez — DISMISSED** [review]

**Review by arielmelendez — APPROVED** [review]

File: src/commands/download_file.ts
Comment (matibat) [inline src/commands/download_file.ts:15]: Windows won't be happy with that extra `/`
```suggestion
		const destOutputPath = parameters.getParameterValue(LocalFilePathDownloadParameter) || '.';
```

**Review by arielmelendez — APPROVED** [review]

**Review by fedellen — APPROVED** [review]

**Review by arielmelendez — APPROVED** [review]

**Review by arielmelendez — DISMISSED** [review]

**Review by fedellen — APPROVED** [review]

**Pull request merged**
